File: src/main/java/rocks/sakira/flamingo/Flamingo.java
package rocks.sakira.flamingo;

import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.entity.SpriteRenderer;
import net.minecraft.entity.EntityClassification;
import net.minecraft.entity.EntitySpawnPlacementRegistry;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.gen.Heightmap;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.client.event.ColorHandlerEvent;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.client.registry.RenderingRegistry;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.ForgeRegistries;
import rocks.sakira.flamingo.client.render.RenderFlamingo;
import rocks.sakira.flamingo.entity.EntityFlamingo;
import rocks.sakira.flamingo.register.Entities;
import rocks.sakira.flamingo.register.Items;
import rocks.sakira.flamingo.register.SoundEvents;

@Mod(Flamingo.MOD_ID)
public class Flamingo {
    public static final String MOD_ID = "flamingo_ooo";

    public Flamingo() {
        final IEventBus eventBus = FMLJavaModLoadingContext.get().getModEventBus();

        eventBus.register(this);

        Entities.REGISTER.register(eventBus);
        Items.REGISTER.register(eventBus);
        SoundEvents.REGISTER.register(eventBus);
    }

    @SubscribeEvent
    public void setupCommon(final FMLCommonSetupEvent event) {
        EntitySpawnPlacementRegistry.register(
                Entities.FLAMINGO_ENTITY.get(),
                EntitySpawnPlacementRegistry.PlacementType.ON_GROUND,
                Heightmap.Type.MOTION_BLOCKING_NO_LEAVES,
                EntityFlamingo::canAnimalSpawn
        );

        for (Biome biome : ForgeRegistries.BIOMES.getValues()) {
            if (biome.getCategory() == Biome.Category.SWAMP) {
                biome.getSpawns(EntityClassification.CREATURE).add(
                        new Biome.SpawnListEntry(
                                Entities.FLAMINGO_ENTITY.get(),
                                15,
                                5,
                                12
                        )
                );
            }
        }
    }

    @SubscribeEvent
    @OnlyIn(Dist.CLIENT)
    public void setupClient(final FMLClientSetupEvent event) {
        RenderingRegistry.registerEntityRenderingHandler(
                Entities.FLAMINGO_ENTITY.get(),
                RenderFlamingo::new
        );

        RenderingRegistry.registerEntityRenderingHandler(
                Entities.FLAMINGO_EGG_ENTITY.get(),
                (rendererManager) -> new SpriteRenderer<>(rendererManager, Minecraft.getInstance().getItemRenderer())
        );
    }


    @SubscribeEvent
    @OnlyIn(Dist.CLIENT)
    public void setupItemColours(final ColorHandlerEvent.Item event) {
        event.getItemColors().register((itemColor, itemsIn) ->
                        Items.FLAMINGO_SPAWN_EGG.get().getColor(itemsIn),
                Items.FLAMINGO_SPAWN_EGG.get()
        );
    }

    @SubscribeEvent
    @OnlyIn(Dist.CLIENT)
    public void clientSetup(final FMLClientSetupEvent event) throws NoSuchFieldException, IllegalAccessException {

    }
}

File: src/main/kotlin/rocks/sakira/flamingo/Flamingo.kt
package rocks.sakira.flamingo

import net.fabricmc.fabric.api.client.rendereregistry.v1.EntityRendererRegistry
import net.fabricmc.fabric.api.loot.v1.event.LootTableLoadingCallback
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.entity.FlyingItemEntityRenderer
import net.minecraft.entity.SpawnGroup
import net.minecraft.entity.SpawnRestriction
import net.minecraft.entity.passive.AnimalEntity
import net.minecraft.entity.projectile.thrown.EggEntity
import net.minecraft.loot.LootTables
import net.minecraft.loot.entry.ItemEntry
import net.minecraft.util.registry.Registry
import net.minecraft.world.Heightmap
import net.minecraft.world.biome.Biome
import rocks.sakira.flamingo.client.render.FlamingoRenderer
import rocks.sakira.flamingo.mixin.LootPoolMixin
import rocks.sakira.flamingo.mixin.LootTableMixin
import rocks.sakira.flamingo.mixin.SpawnRestrictionMixin
import rocks.sakira.flamingo.register.Effects
import rocks.sakira.flamingo.register.Entities
import rocks.sakira.flamingo.register.Items
import rocks.sakira.flamingo.register.Sounds

const val MOD_ID = "flamingo_ooo"

fun init() {
    Items.register()
    Sounds.register()
    Effects.register()
    Entities.registerAttributes()

    registerSpawnRestrictions()
    registerSpawns()
    registerShrimpLootEntry()
}

fun clientInit() {
    EntityRendererRegistry.INSTANCE.register(Entities.FLAMINGO) { dispatcher, _ ->
        FlamingoRenderer(dispatcher)
    }

    EntityRendererRegistry.INSTANCE.register(Entities.FLAMINGO_EGG) { dispatcher, _ ->
        FlyingItemEntityRenderer<EggEntity>(dispatcher, MinecraftClient.getInstance().itemRenderer)
    }
}


fun registerSpawnRestrictions() {
    SpawnRestrictionMixin.register(
        Entities.FLAMINGO,
        SpawnRestriction.Location.ON_GROUND,
        Heightmap.Type.MOTION_BLOCKING_NO_LEAVES,
        AnimalEntity::isValidNaturalSpawn
    )
}

fun registerSpawns() {
    for (biome in Registry.BIOME.stream()) {
        if (biome.category == Biome.Category.SWAMP) {
            biome.getEntitySpawnList(SpawnGroup.CREATURE).add(
                Biome.SpawnEntry(
                    Entities.FLAMINGO,
                    15,
                    5,
                    12
                )
            )
        }
    }
}

fun registerShrimpLootEntry() {
    LootTableLoadingCallback.EVENT.register(LootTableLoadingCallback { resourceManager, manager, id, supplier, setter ->
        if (id == LootTables.FISHING_FISH_GAMEPLAY) {
            val table = manager.getTable(id)
            val pool = (table as LootTableMixin).pools.first()
            val entries = (pool as LootPoolMixin).entries.toMutableList()

            entries.add(
                ItemEntry.builder(Items.SHRIMP)
                    .weight(13)
                    .build()
            )

            (pool as LootPoolMixin).entries = entries.toTypedArray()
        }
    })
}
